src/util-plugin.c

```diff
@@ -0,0 +1,178 @@
+/* Copyright (C) 2020 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+#include "suricata-common.h"
+#include "suricata-plugin.h"
+#include "util-plugin.h"
+
+#ifdef HAVE_PLUGINS
+
+#include <dlfcn.h>
+
+static TAILQ_HEAD(, SCPluginFileType_) output_types =
+    TAILQ_HEAD_INITIALIZER(output_types);
+
+static TAILQ_HEAD(, SourcePlugin_) source_plugins = TAILQ_HEAD_INITIALIZER(source_plugins);
+
+static void InitPlugin(char *path)
+{
+    void *lib = dlopen(path, RTLD_NOW);
+    if (lib == NULL) {
+        SCLogNotice("Failed to open %s as a plugin: %s", path, dlerror());
+    } else {
+        SCLogNotice("Loading plugin %s", path);
+        SCPlugin *plugin = dlsym(lib, "PluginSpec");
+        if (plugin == NULL) {
+            SCLogError(SC_ERR_PLUGIN, "Plugin does not export a plugin specification: %s", path);
```

Comment:
should this be fatal?

Reply:
I don't see it as  fatal.  The plugin will fail to load tho. It would most likely be a result of a non-plugin DSO being loaded.